### PR TITLE
Features/multiple parallel heliostats

### LINKED ISSUE #90: Multi Heliostat Support - Data Parallel, Heliostat Raytracing
`ARTIST` must support multiple heliostats; currently, only one heliostat is supported.

**Describe the solution you'd like**
To initially enable multiple heliostat support, we will implement this in the following manner:

- [ ] Heliostat-based raytracing.
- [ ] Data parallel (no model parallel).

This may have some limitations, but it should enable the simulation of small power plants, which is sufficient for initial use cases.

**Planned Deadline:** November 15, 2024

### PR DESCRIPTION
**Pull request by MarleneBusch**
# Description

I have adapted the datastructures in ARTIST to enable parallel processing of multiple heliostats. Alignment and raytracing can now run simultaneously for thousands of heliostats on a single device. Additionally it is possible to distribute heliostat groups onto multiple devices with Distributed Data Parallel. I had to change almost all files because of the new datastructures.
I have also added some lines to the nurbs code to accelerate this part of the code. Since we never actually use non-uniform splines we can simplify some methods and differentiate between uniform and non-uniform splines now.

Fixes #90  #98 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: tutorials/01_heliostat_raytracing_tutorial.py [hunk 1]
```diff
@@ -9,20 +9,20 @@
...
 
 # If you have already generated the tutorial scenario yourself, you can leave this boolean as False. If not, set it to
 # true and a pre-generated scenario file will be downloaded for this tutorial!
-DOWNLOAD_DATA = False
-scenario_file = pathlib.Path("please/insert/the/path/to/the/scenario/here/name.h5")
+default = False
+scenario_path = pathlib.Path("please/insert/the/path/to/the/scenario/here/scenario.h5")
 
-if DOWNLOAD_DATA:
+if default:
```
Comment (kalebphipps) [inline tutorials/01_heliostat_raytracing_tutorial.py:20]: Is this link still ok or do we need to update the online scenario?

Or maybe just remove it entirely and keep one scenario saved in git?

**Review by kalebphipps — APPROVED** [review]
Thank you for all the changes!

It looks good to me - I have a few minor questions, but I am happy for you to merge this without me reviewing it again!

File: artist/field/actuators.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    """
+
+    def __init__(
+        self,
+    ) -> None:
+        """
+        Initialize abstract actuators.
+
+        The abstract actuator implements a template for the construction of inheriting actuators which can be
+        ideal or linear. An actuator is responsible for turning the heliostat surface in such a way that the
```
[the remark this answers is not in the document]
Comment (MarleneBusch) [reply artist/field/actuators.py:27]: I removed the specific actuator types.

File: artist/raytracing/heliostat_tracing.py [hunk 3]
```diff
@@ -498,109 +491,94 @@ def sample_bitmap(
         # That the "." may be anywhere in-between the four pixels is not
         # shown in the ASCII diagram, but is important to keep in mind.
 
-        # The lower-valued neighboring pixels (for x this corresponds to 4
-        # and 3, for y to 1 and 4).
-        x_inds_low = x_ints.floor().long()
-        y_inds_low = y_ints.floor().long()
+        # The lower-valued neighboring pixels (for x this corresponds to 1
+        # and 4, for y to 3 and 4).
```
[the remark this answers is not in the document]
Comment (MarleneBusch) [reply artist/raytracing/heliostat_tracing.py:495]: As I have looked at those pieces of the code extensively, the docstrings in this part of the code seem very obvious to me, however I can understand that this is not the case for everyone. 
Maybe you can help me make the Docstrings more precise and understandable.

In the comments in line 484 we have the ASCII diagram:
        # 1|2
        # -.-
        # 4|3
"The lower-valued neighboring pixels" means: A ray wont hit the intersection of 4 pixels exactly. It will be nearest to one pixel and farthest away from another. This happens in two dimensions x and y. Looking at only the x-dimension. Lower valued pixels are 1 and 4 while a higher x value means it is either at 2 or 3. The same consideration happens along the y-axis. 
As far as I understand the numbering of the pixels into the ASCII-diagram
        # 1|2
        # -.-
        # 4|3
is not part of any literature and can be numbered in any way that we want, we just chose this way because it numbers the pixels clockwise. Any numbering would be okay if the variables in the script are adapted accordingly.

Do you have a suggestion for better docstrings or a new numbering system altogether?

File: artist/scene/light_source.py [hunk 4]
```diff
@@ -45,17 +45,22 @@ def __init__(
...
+        device: Union[torch.device, str] = "cuda",
     ) -> Self:
         """
         Load the light source from an HDF5 file.
 
         Parameters
         ----------
         config_file : h5py.File
-            The HDF5 file containing the information about the light source.
+            The HDF5 file containing the information about the sun.
```
[the remark this answers is not in the document]
Comment (MarleneBusch) [reply artist/scene/light_source.py:59]: We still have the option to implement artificial light sources in the future.
Your suggestion makes sense, i updated the docstring.

File: tutorials/01_heliostat_raytracing_tutorial.py (hunk 1 above)
Comment (MarleneBusch) [reply tutorials/01_heliostat_raytracing_tutorial.py:20]: I think we should remove the link and provide a scenario in git. I have changed the code accordingly.

File: artist/field/heliostat_field.py [hunk 5]
```diff
@@ -66,36 +153,301 @@ def from_hdf5(
...
+            (number_of_heliostats, number_of_surface_points_per_heliostat, 4),
+            device=device,
+        )
+        all_surface_normals = torch.zeros(
+            (number_of_heliostats, number_of_surface_points_per_heliostat, 4),
+            device=device,
+        )
+        all_initial_orientations = torch.zeros((number_of_heliostats, 4), device=device)
+
+        all_kinematic_deviation_parameters = torch.zeros(
```
[the remark this answers is not in the document]
Comment (MarleneBusch) [reply artist/field/heliostat_field.py:198]: done

File: artist/util/config_dictionary.py [hunk 6]
```diff
@@ -95,14 +93,16 @@
...
 
-kinematic_initial_orientation = "initial_orientation"
-"""Key to access the initial orientation."""
 kinematic_type = "type"
 """Key to access the kinematic type."""
 rigid_body_key = "rigid_body"
 """Key to define that the rigid body kinematic is used as a kinematic type."""
-kinematic_offsets_key = "offsets"
-"""Key to access the kinematic offsets."""
+rigid_body_number_of_deviation_parameters = 18
```
[the remark this answers is not in the document]
Comment (MarleneBusch) [reply artist/util/config_dictionary.py:100]: done

File: artist/raytracing/heliostat_tracing.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (MarleneBusch) [reply artist/raytracing/heliostat_tracing.py:495]: Thats okay 😊 I will leave it as it is

File: artist/field/heliostat_field.py [hunk 7]
```diff
@@ -66,36 +153,301 @@ def from_hdf5(
...
+
+        rigid_body_kinematic = RigidBody(
+            number_of_heliostats=self.number_of_heliostats,
+            heliostat_positions=self.all_heliostat_positions,
+            aim_points=self.all_aim_points,
+            actuator_parameters=self.all_actuator_parameters,
+            initial_orientations=self.all_initial_orientations,
+            deviation_parameters=self.all_kinematic_deviation_parameters,
+            device=device,
+        )
```
[the remark this answers is not in the document]
Comment (MarleneBusch) [reply artist/field/heliostat_field.py:439]: will be solved when we allow different heliostat groups with different kinematics.
